node/src/components/diagnostics_port/tasks.rs

```diff
@@ -0,0 +1,630 @@
+/// indicating the outcome of the operation.
+#[derive(Debug, Serialize)]
+enum Outcome {
+    /// Operation succeeded.
+    Success {
+        /// Human-readable message giving additional info and/or stating the effect.
+        msg: String,
+    },
+    /// Operation failed.
+    Failure {
+        /// Human-readable message describing the failure that occurred.
+        reason: String,
+    },
+}
+
+impl Outcome {
+    /// Constructs a new successful outcome.
+    fn success<S: ToString>(msg: S) -> Self {
+        Outcome::Success {
+            msg: msg.to_string(),
+        }
+    }
+
+    /// Constructs a new failed outcome.
+    fn failed<S: ToString>(reason: S) -> Self {
+        Outcome::Failure {
+            reason: reason.to_string(),
+        }
+    }
+}
+
+impl Display for Outcome {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Outcome::Success { msg } => {
+                write!(f, "OK   {}", msg)
+            }
+            Outcome::Failure { reason } => {
+                write!(f, "ERR  {}", reason)
+            }
+        }
+    }
+}
+
+/// Configuration for a connection diagnostics port session.
+#[derive(Copy, Clone, Debug, Default, Serialize)]
+struct Session {
+    /// Whether or not to suppress the operation outcome.
+    quiet: bool,
+    /// Output format to send to client.
+    output: OutputFormat,
+}
+
+impl Display for Session {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+/// A serializer supporting multiple format variants that writes into a file.
+pub enum FileSerializer {
+    /// JSON-format serializer.
+    Json(serde_json::Serializer<File>),
+    /// Bincode-format serializer.
+    Bincode(
+        bincode::Serializer<
+            File,
+            WithOtherTrailing<WithOtherIntEncoding<DefaultOptions, FixintEncoding>, AllowTrailing>,
+        >,
+    ),
+}
+
+impl FileSerializer {
+    /// Converts the temp file serializer into an actual erased serializer.
+    pub fn as_serializer<'a>(&'a mut self) -> Box<dyn ErasedSerializer + 'a> {
+        match self {
+            FileSerializer::Json(json) => Box::new(<dyn erased_serde::Serializer>::erase(json)),
+            FileSerializer::Bincode(bincode) => {
+                Box::new(<dyn erased_serde::Serializer>::erase(bincode))
+            }
+        }
+    }
+}
+
+/// Error obtaining a queue dump.
+#[derive(Debug, Error)]
+enum ObtainDumpError {
+    /// Error trying to create a temporary directory.
+    #[error("could not create temporary directory")]
+    CreateTempDir(#[source] io::Error),
+    /// Error trying to create a file in the temporary directory.
+    #[error("could not create file in temporary directory")]
+    CreateTempFile(#[source] io::Error),
+    /// Error trying to reopen the file in the temporary directory after writing.
+    #[error("could not reopen file in temporary directory")]
+    ReopenTempFile(#[source] io::Error),
```

Comment:
Where is this `#[source]` coming from? What crate? Will it be included in the error output?

Reply:
Its from `thiserror::Error`. It will generate a [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) implementation (the default is to return `None`), which is just good practice.

There are basically two "styles" of chaining errors - one is to use the properly typed methods like `source()` (see link above), which will show up cleanly in an error trace, the other is just stuffing it into a string, e.g.

```rust
 #[error("could not reopen file in temporary directory: {0}")]
    ReopenTempFile(io::Error),
```

I prefer the former, the only drawback is that the default `Display` implementation generated by `thiserror` does not automatically add the `: {}`, as it may be used in an explicit backtrace-style printing (see the docs on `source()` above, contents of the main function).

I usually try to get the best of both worlds by using our helper function `display_error`.

Reply:
I like this one more as well. Thanks.